app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourcesIndexRepo.kt

```diff
@@ -59,31 +62,28 @@ class ResourcesIndexRepo(
     }
 
     suspend fun provide(rootAndFav: RootAndFav): ResourcesIndex = withContext(Dispatchers.IO) {
-        val roots = resolveRootAndFav(rootAndFav)
-        val indexShards = roots.map { root ->
-            indexByRoot[root] ?: let {
-                val index = loadFromDatabase(root)
-                indexByRoot[root] = index
-                index
+        provideMutex.withLock {
+            val roots = foldersRepo.resolveRoots(rootAndFav)
+            val indexShards = roots.map { root ->
+                indexByRoot[root] ?: let {
+                    val index = loadFromDatabase(root)
```

Comment:
I think, we can use `foldersRepo.resolveRoots(..)` without locking.
It has lock inside and we don't need locking for local variable `roots`.

About `indexByRoot` map: if we can use "read-write lock" instead of mutex, then we would do read-lock when we iterate its entries (no real locking if there is not write), and only do write-lock when we have no index loaded in particular entry. This way, in case we have all roots loaded, we don't write anything new into `indexByRoot`, so we do only read-lock which is much cheaper.